reps/2023-07-28-serve-grpc-proxy.md

```diff
@@ -0,0 +1,501 @@
+    response as `ServeResponse` object.
+- `HTTPProxyActor` will be changed to start both `HTTPProxy` and `gRPCProxy`.
+  `gRPCProxy` is only setup when the `gRPCOptions` is provided in the config.
+- `LongestPrefixRouter` to add a new method `match_target()` to help gRPC look up for
+  the routes from the application name
+
+#### Handle and Router
+New flag `serve_grpc_request` is added to the `HandleOptions` and to the
+`RequestMetadata` to signal to the replicas when it’s a gRPC request. This flag
+is boolean defaulting to false. It's only set when `gRPCProxy` is setting up the
+handle.
+
+#### Replica
+- `call_user_method_grpc_unary()` is added to handle unary gRPC requests
+- `call_user_method_with_grpc_streaming()` is added to handle streaming gRPC requests
+- A new if-statement is added to use gRPC methods when the `serve_grpc_request`
+  flag is set true in `handle_request()` and in `handle_request_streaming()`
+
+#### Data Models
+- `gRPCRequest` is added to replace `StreamingHTTPRequest` in `RayServeHandle`
+  and `Router` to be used for gRPC requests. It contains a `grpc_user_request`
+  field to store the byte input for the replica and a `grpc_proxy_handle` field
+  to store the ray serve handle to be used for gRPC requests.
+- `ServeRequest` is a new data model served as the input to `proxy_request()`
+  method so any types of proxy can share the same GA entry point.
+- `ASGIServeRequest` is a subclass of `ServeRequest` used by `HTTPProxy`.
+  It contains the implementations to extract attributes from ASGI input objects for
+  sending requests to the replica
+- `gRPCServeRequest` is a subclass of `ServeRequest` used by `gRPCProxy`.
+  It contains the implementations to extract attributes from gRPC input objects
+  for sending requests to the replica
+- `ServeResponse` is a new data model served as the output of `proxy_request()`
+  method. It has required the string `status_code` used for metrics tracking. It also
+  has optional `response` and `streaming_response` allowing gRPC to pass response
+  back to the client
+
+### gRPCOptions
+This is a new config option we added to the serve config. It contains two fields:
+- `port`: TPort for gRPC server. Will only start gRPC server if set. Cannot be
+  updated once Serve has started running. Serve must be shut down and restarted
+  with the new port instead.
+- `grpc_servicer_functions`: The servicer functions used to add the method handlers
+  to the gRPC server. Default to empty list, which means no gRPC methods will be
+  added and no gRPC server will be started. The servicer functions need to be
+  importable from the context of where Serve is running.
+
+If Serve is started with `gRPCOptions` provided, the `HTTPProxyActor` will start
+the gRPC proxy with the provided port and the provided servicer functions.
+
+#### gRPC Util
+This is where our implementation of gRPC the serialization and deserialization lives.
```

Comment:
That's a good point. Is there any reason why we cannot defer deserialization until the request hits the replica itself? That would seem to simplify the gRPC support considerably.

Reply:
Gene and I looked into a number of ways to do this and it's unfortunately non-trivial. There doesn't appear to be a clean way to do it without modifying client code while still using gRPC's server implementation. That means we'd need to run a raw HTTP/2 server that's fully compatible with gRPC clients. This may be doable, but it introduces a lot more risk/uncertainty.

Reply:
Agreed. I think right now the biggest hurdle is feature completeness and increase adoption. We can extend this in the future :)

Reply:
Agreed. We shouldn't block on this initially, we can do it as a follow up.